sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -122,8 +121,11 @@ class BatchLoads<DestinationT, ElementT>
   static final int FILE_TRIGGERING_RECORD_COUNT = 500000;
   // If user triggering is supplied, we will trigger the file write after this many bytes are
   // written.
+  private static final int UPLOAD_CHUNK_SIZE_GRANULARITY = 8 * 1024 * 1024;
   static final int DEFAULT_FILE_TRIGGERING_BYTE_COUNT =
-      AsyncWriteChannelOptions.UPLOAD_CHUNK_SIZE_DEFAULT; // 64MiB as of now
+      Runtime.getRuntime().maxMemory() < 512 * 1024 * 1024
```

Comment:
this variable was made private in gcs-connector 3.x, so I just copied the value over. https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v3.0.1/util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java#L72